src/transpile/fusion.hpp

```diff
@@ -84,57 +683,52 @@ class Fusion : public CircuitOptimization {
-                        Method method) const;
-
-  double get_cost(const op_t& op) const;
-
   void optimize_circuit(Circuit& circ,
-                        Noise::NoiseModel& noise,
+                        const Noise::NoiseModel& noise,
                         const opset_t &allowed_opset,
-                        uint_t ops_start,
-                        uint_t ops_end) const;
-
-  bool aggregate_operations(oplist_t& ops,
-                            const int fusion_start,
-                            const int fusion_end,
-                            uint_t max_fused_qubits,
-                            Method method) const;
-
-  // Aggregate a subcircuit of operations into a single operation
-  op_t generate_fusion_operation(const std::vector<op_t>& fusioned_ops,
-                                 const reg_t &num_qubits,
-                                 Method method) const;
-
-  bool is_diagonal(const oplist_t& ops,
-                   const uint_t from,
-                   const uint_t until) const;
-
-  double estimate_cost(const oplist_t& ops,
-                       const uint_t from,
-                       const uint_t until) const;
-
-  void add_fusion_qubits(reg_t& fusion_qubits, const op_t& op) const;
+                        const uint_t ops_start,
+                        const uint_t ops_end,
+                        const std::shared_ptr<Fuser> fuser,
+                        const FusionMethod& method) const;
 
 #ifdef DEBUG
-  void dump(const Circuit& circuit) const;
+  void dump(const Circuit& circuit) const {
+    auto& ops = circuit.ops;
+    for (uint_t op_idx = 0; op_idx < ops.size(); ++op_idx) {
+      std::cout << std::setw(3) << op_idx << ": ";
+      if (ops[op_idx].type == optype_t::nop) {
+        std::cout << std::setw(15) << "nop" << ": ";
+      } else {
+        std::cout << std::setw(15) << ops[op_idx].name << "-" << ops[op_idx].qubits.size() << ": ";
+        if (ops[op_idx].qubits.size() > 0) {
+          auto qubits = ops[op_idx].qubits;
+          std::sort(qubits.begin(), qubits.end());
+          int pos = 0;
+          for (int j = 0; j < qubits.size(); ++j) {
+            int q_pos = 1 + qubits[j] * 2;
+            for (int k = 0; k < (q_pos - pos); ++k) {
+              std::cout << " ";
+            }
+            pos = q_pos + 1;
+            std::cout << "X";
+          }
+        }
+      }
+      std::cout << std::endl;
+    }
+  }
 #endif
 
 private:
-  const static Operations::OpSet noise_opset_;
+  std::vector<std::shared_ptr<Fuser>> fusers;
 };
 
-
-const Operations::OpSet Fusion::noise_opset_(
-  {Operations::OpType::kraus,
-   Operations::OpType::superop,
-   Operations::OpType::reset},
-  {}, {}
-);
-
+Fusion::Fusion() {
+  fusers.push_back(std::make_shared<DiagonalFusion>());
+  fusers.push_back(std::make_shared<NQubitFusion<1>>());
+  fusers.push_back(std::make_shared<NQubitFusion<2>>());
+  fusers.push_back(std::make_shared<CostBasedFusion>());
```

Comment:
Is there a usecase where we wouldn't want to use all these `fusers` but only a few them (selected by conf for example)?

Reply:
When performance degradation happens because of fusion, we will disable some of them for diagnostics. All of fusers can be disabled with configuration (such as `fusion_enable.diagonal`).

Reply:
ok, you always have the fusers and you can disable by config.